pubsub/schema.go

```diff
@@ -0,0 +1,292 @@
+		return nil, err
+	}
+	return protoToSchemaConfig(s), nil
+}
+
+// Schemas returns an iterator which returns all of the schemas for the client's project.
+//
+// It is EXPERIMENTAL and subject to change or removal without notice.
+func (c *SchemaClient) Schemas(ctx context.Context, view SchemaView) *SchemaIterator {
+	return &SchemaIterator{
+		it: c.sc.ListSchemas(ctx, &pb.ListSchemasRequest{
+			Parent: fmt.Sprintf("projects/%s", c.projectID),
+			View:   pb.SchemaView(view),
+		}),
+	}
+}
+
+// SchemaIterator is a struct used to iterate over schemas.
+//
+// It is EXPERIMENTAL and subject to change or removal without notice.
+type SchemaIterator struct {
+	it  *vkit.SchemaIterator
+	err error
+}
+
+// Next returns the next schema. If there are no more schemas, iterator.Done will be returned.
+func (s *SchemaIterator) Next() (*SchemaConfig, error) {
+	if s.err != nil {
+		return nil, s.err
+	}
+	pbs, err := s.it.Next()
+	if err != nil {
+		return nil, err
+	}
+	return protoToSchemaConfig(pbs), nil
+}
+
+func (s *SchemaClient) DeleteSchema(ctx context.Context, schemaPath string) error {
+	return s.sc.DeleteSchema(ctx, &pb.DeleteSchemaRequest{
+		Name: schemaPath,
+	})
+}
+
+// ValidateSchemaResult is the response for the ValidateSchema method.
+// Reserved for future use.
+//
+// It is EXPERIMENTAL and subject to change or removal without notice.
+type ValidateSchemaResult struct{}
+
+// ValidateSchema validates a schema.
+//
+// It is EXPERIMENTAL and subject to change or removal without notice.
+func (s *SchemaClient) ValidateSchema(ctx context.Context, schema SchemaConfig) (*ValidateSchemaResult, error) {
+	req := &pb.ValidateSchemaRequest{
+		Parent: fmt.Sprintf("projects/%s", s.projectID),
+		Schema: schema.toProto(),
+	}
+	_, err := s.sc.ValidateSchema(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	return &ValidateSchemaResult{}, nil
+}
+
+// ValidateMessageResult is the response for the ValidateMessage method.
+// Reserved for future use.
+//
+// It is EXPERIMENTAL and subject to change or removal without notice.
+type ValidateMessageResult struct{}
+
+// ValidateMessage validates a message against an schema specified
+// by a schema config.
+func (s *SchemaClient) ValidateMessageWithConfig(ctx context.Context, msg []byte, encoding SchemaEncoding, config SchemaConfig) (*ValidateMessageResult, error) {
+	req := &pb.ValidateMessageRequest{
+		Parent: fmt.Sprintf("projects/%s", s.projectID),
+		SchemaSpec: &pb.ValidateMessageRequest_Schema{
+			Schema: config.toProto(),
+		},
+		Message:  msg,
+		Encoding: pb.Encoding(encoding),
+	}
+	_, err := s.sc.ValidateMessage(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	return &ValidateMessageResult{}, nil
+}
+
+// ValidateMessage validates a message against an schema specified
+// by a schema path pointing to an existing schema.
+func (s *SchemaClient) ValidateMessageWithPath(ctx context.Context, msg []byte, encoding SchemaEncoding, schemaPath string) (*ValidateMessageResult, error) {
```

Comment:
Yeah I have mixed feelings about this. We might want to allow users to access schemas outside of their project. This is sort of true for Pub/Sub as well, so we provide additional methods like `TopicInProject` and `SubscriptionInProject`, but I don't particularly like this overloaded concept. I think if we clarify the format for schemaPath when we use this (in this method as well as in the get `Schema`, this should be ok as a path.

Reply:
@kamalaboulhosn proposed a solution which is to default to `schemaIDs` rather than path, but not provide a `SchemaInProject`. Instead, users can instantiate another `schemaClient` if they want to access schemas outside of their existing project. I like this too, so I'll implement this for now.